doc: add docs for dealing with special characters in ref names

wincent opened the pull request:
### Why:

There are a few places in the UI where we provide instructions for how to do things on the command-line. We can provide shell-friendly escaped versions of _most_ branch names, but there will always be some edge-casey ones (eg. names containing single quotes) for which the escaping rules vary subtly by shell. Rather than try to cram a dozen variants of the commands into the UI, we want to direct people to the docs for such problematic names.

### What's being changed:

The intent here is not to document every shell's rules (doomed to be incomplete and become stale over time), but rather provide a flavour of what users should be aware of, and some helpful links to jumping off to official upstream docs for commonly-used shells.

Closes: https://github.com/github/coding/issues/2388

<img width="1268" alt="Google Chrome app 2021-09-06 17 14 02 UTC" src="https://user-images.githubusercontent.com/7074/132248162-c044024c-daea-4816-a8cd-e255bd8e0c86.png">

### Check off the following:

- [x] I have reviewed my changes in staging (look for the latest deployment event in your pull request's timeline, then click **View deployment**).
- [x] For content changes, I have completed the [self-review checklist](https://github.com/github/docs/blob/main/CONTRIBUTING.md#self-review).

### Writer impact (This section is for GitHub staff members only):

- [ ] This pull request impacts the contribution experience
  - [ ] I have added the 'writer impact' label
  - [ ] I have added a description and/or a video demo of the changes below (e.g. a "before and after video")

<!-- Description of the writer impact here -->

cc @codeminator and @georgebrock

wincent committed:
docs: fix `shortTitle` copy-pasta mistake

wincent committed:
test: rename file to match document title

Because it was failing this test:

    ● pages module › loadPages › every English page has a filename that matches its slugified title

as seen here:

    https://github.com/github/docs/pull/9577/checks?check_run_id=3524790084

wincent committed:
docs: edit for simplicity

As per style guide:

    https://github.com/github/docs/blob/main/contributing/content-style-guide.md

and self-review checklist:

    https://github.com/github/docs/blob/main/CONTRIBUTING.md#self-review

- Avoided nominalization in one place (eg. "for structure" instead of
  "to provide organization", which might have been confused with a
  _GitHub_ organization).
- Change passive to active voice in one place (eg. "most shells would
  interpret" instead of "would be interpreted by most shells").

georgebrock commented on content/get-started/using-git/dealing-with-special-characters-in-branch-and-tag-names.md:
```diff
@@ -0,0 +1,25 @@
+  ghae: '*'
+shortTitle: Special characters in names
+---
+Most repositories use simple branch names like `main` or `update-icons`. Likewise, tag names usually follow a basic format, such as a version number like `v1.2.3`. Both branch and tag names may also use the path separator (`/`) for structure (for example, `area/item`, or `level-1/level-2/level-3`). Apart from a handful of exceptions &mdash; such as not starting or ending a name with a slash, or h ...
```
Small accessibility suggestion: can we expand the link text from “very few restrictions” to a larger part of the sentence, something like “very few restrictions on what characters may be used in branch and tag names.” Ideally it would make sense in isolation, e.g. when using a screen reader in links mode.

(see [WCAG 2.1, success criterion 2.4.4: Link Purpose (In Context)][2.4.4] and [WCAG 2.1, success criterion 2.4.9: Link Purpose (Link Only)][2.4.9])

[2.4.4]: https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html
[2.4.9]: https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-link-only.html

wincent replied:
Good call. Thanks @georgebrock!

georgebrock commented on content/get-started/using-git/dealing-with-special-characters-in-branch-and-tag-names.md:
```diff
@@ -0,0 +1,25 @@
+* [Zsh](https://www.zsh.org/) behaves similarly to Bash, in that single quotes turn off the special behavior of most characters inside them (although, note that this behavior is configurable via the `RC_QUOTES` option).
+* [PowerShell](https://microsoft.com/powershell) likewise treats characters verbatim when inside single quotes.
+
+As such, most branch and tag names with special characters can be handled by wrapping the name in single quotes. The main exception to this rule is when the branch name itself contains a single quote. For edge cases like this, it is best to consult the official documentation provided with your shell. Starting points for some common shells include:
```
What do you think about moving the first sentence of this paragraph before the list? It seems like the list could use a little more introduction to connect it to the idea of quoting or escaping appropriately in the prior paragraph.

Something like:

> Most branch and tag names with special characters can be quoted by wrapping the name in single quotes:
>
> - In the Bash shell, …
>
> The main exception to this rule is …

wincent replied:
Sounds good @georgebrock. 👍

georgebrock reviewed (commented):
These are great docs! I made a couple of very small suggestions, but this is already very thorough and helpful as-is ✨

wincent committed:
docs: include more text in link text for accessibility

As suggested here:

    https://github.com/github/docs/pull/9577/files#r703346187

Links:

- https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-in-context.html
- https://www.w3.org/WAI/WCAG21/Understanding/link-purpose-link-only.html

wincent committed:
docs: move sentence higher up in document

As suggested here:

    https://github.com/github/docs/pull/9577/files#r703350737

wincent committed:
docs: add docs for dealing with special characters in ref names

There are a few places in the UI where we provide instructions for how
to do things on the command-line. We can provide shell-friendly escaped
versions of _most_ branch names, but there will always be some
edge-casey ones (eg. names containing single quotes) for which the
escaping rules vary subtly by shell. Rather than try to cram a dozen
variants of the commands into the UI, we want to direct people to the
docs for such problematic names.

The intent here is not to document every shell's rules (doomed to be
incomplete and become stale over time), but rather provide a flavour of
what users should be aware of, and some helpful links to jumping off to
official upstream docs for commonly-used shells.

Closes: https://github.com/github/coding/issues/2388

ramyaparimi commented:
@wincent 
Thanks so much for opening a PR! I'll get this triaged for review :zap:

b4mboo committed:
Merge branch 'main' into wincent/special-characters

codeminator commented:
@ramyaparimi any thoughts on when this PR will be reviewed?

ramyaparimi commented:
@codeminator Thanks for reaching out.

We have a limited set of resources to pull from in order to get work done in the open source repository. We do our best to get things addressed as quickly as possible. Unfortunately, we can’t promise exactly when this will be handled, but rest assured that it is in the hands of the proper team and we’ll get to it as rapidly as we can with resources and other priorities in mind.

Thanks for your patience and your contributions to the GitHub docs.

lucascosti committed:
Docs team edit

lucascosti committed:
Merge branch 'main' into wincent/special-characters

lucascosti commented:
👋 Thanks a lot for providing the content for this article, @wincent! 🙇‍♂️ I'm sorry that it's taken us a while to get to reviewing this.

I've push up an edit in https://github.com/github/docs/pull/9577/commits/1b903fcff36197e524c9283b39f0a3dab0986e1f, which breaks the content up into digestible sections, and also made some wording/formatting edits for it to align with our style guides and language recommendations.

Please let me know if I've unintentionally changed the meaning of anything that makes it incorrect.

Because I've made significant changes, I'll get someone else on the team to give it a quick review of my changes.

martin389 commented on content/get-started/using-git/dealing-with-special-characters-in-branch-and-tag-names.md:
```diff
@@ -0,0 +1,48 @@
+
+## Naming branches and tags
+
+If possible, create branch and tag names that don't contain special characters which must be escaped. A safe default set of characters to use for branch names and tag names is:
```
```suggestion
If possible, create branch and tag names that don't contain special characters, as these would need to be escaped. A safe default set of characters to use for branch names and tag names is:
```

martin389 commented on content/get-started/using-git/dealing-with-special-characters-in-branch-and-tag-names.md:
```diff
@@ -0,0 +1,48 @@
+
+Most repositories use simple branch names, such as `main` or `update-icons`. Tag names also usually follow a basic format, such as a version number like `v1.2.3`. Both branch names and tag names may also use the path separator (`/`) for structure, for example `area/item` or `level-1/level-2/level-3`. Other than some exceptions &mdash; such as not starting or ending a name with a slash, or having  ...
+
+## Why escaping special characters is necessary
```
```suggestion
## Why you need to escape special characters
```

martin389 commented on content/get-started/using-git/dealing-with-special-characters-in-branch-and-tag-names.md:
```diff
@@ -0,0 +1,48 @@
+
+## Why escaping special characters is necessary
+
+When using a CLI, you might have situations where a branch or tag name contains special characters that have a special meaning for your shell environment. These characters must be quoted or escaped for you to use them safely in a Git command, otherwise running the command may have unintended effects.
```
```suggestion
When using a CLI, you might have situations where a branch or tag name contains special characters that have a special meaning for your shell environment. To use these characters safely in a Git command, they must be quoted or escaped, otherwise the command may have unintended effects.
```

martin389 commented on content/get-started/using-git/dealing-with-special-characters-in-branch-and-tag-names.md:
```diff
@@ -0,0 +1,48 @@
+
+When using a CLI, you might have situations where a branch or tag name contains special characters that have a special meaning for your shell environment. These characters must be quoted or escaped for you to use them safely in a Git command, otherwise running the command may have unintended effects.
+
+For example, the `$` character is used by many shells to refer to a variable. Most shells would interpret a valid branch name like `hello-$USER` as equivalent to the word "hello", followed by a hyphen, followed by the current value of the `USER` variable, rather than the literal string `hello-$USER`. If a branch name includes the `$` character, then we need to stop the shell from expanding it as  ...
```
```suggestion
For example, the `$` character is used by many shells to refer to a variable. Most shells would interpret a valid branch name like `hello-$USER` as equivalent to the word "hello", followed by a hyphen, followed by the current value of the `USER` variable, rather than the literal string `hello-$USER`. If a branch name includes the `$` character, then the shell must be stopped from expanding it as a variable reference. Similarly, if a branch name contains a semi-colon (`;`), most shells interpret it as a command separator, so it needs to be quoted or escaped.
```

martin389 reviewed (commented):
Thanks for this PR! I've added some small suggestions 👍

lucascosti committed:
Apply suggestions from code review

Co-authored-by: Martin Lopes <martin389@github.com>

lucascosti commented:
Thanks, @martin389! 🙇‍♂️ 

@wincent Let me know if you're happy to get this merged in 🙂

wincent commented:
> Let me know if you're happy to get this merged in 

@lucascosti, thanks for the edits and @martin389, thanks for the review! This looks great to me.

lucascosti committed:
Merge branch 'main' into wincent/special-characters
